### PR TITLE
Migrate to coingecko from cryptocompare

### PR DESCRIPTION
**Pull request by blurpesec**
[ch3535] [ch3536]
### Description
Migrate from cryptocompare to coingecko. Use asset mapping from mycryptohq/assets

Fixes handling token detail fetches for large amounts of assets.
Fixes a bug that results in token detail fetches occurring 8x times for every one time it needs to be run.

### ToDo:
- [x] Fix cors / csp issue with https://price.mycryptoapi.com.

~If you want to test before this csp /cors issue is fixed, find and replace `https://price.mycryptoapi.com` with `https://raw.githubusercontent.com/MyCryptoHQ/assets/master/assets/assets.json`~

### CODE REVIEW

File: common/v2/services/ApiService/TokenInfo/TokenInfo.ts [hunk 1]
```diff
@@ -21,14 +22,37 @@ export default class TokenInfoService {
...
+const spliceContractAddresses = (contractAddresses: string[]): string[][] => {
+  const lengthToSplitOn = 20;
+  const splicedContractAddressArrays = [];
+  do {
+    splicedContractAddressArrays.push([...contractAddresses.splice(0, lengthToSplitOn)]);
+  } while (contractAddresses.length >= lengthToSplitOn);
+
+  splicedContractAddressArrays.push([...contractAddresses]);
+  return splicedContractAddressArrays;
+};
```
Comment (vilsbole) [inline common/v2/services/ApiService/TokenInfo/TokenInfo.ts]: Naming:
This function is not specific to ContractAddresses and will work on any list.
To increase reuse you can add the length as an argument and use a more generic name such as `chunk`
https://medium.com/@Dragonza/four-ways-to-chunk-an-array-e19c889eac4

Ideally use [`R.splitEvery`](https://ramdajs.com/docs/#splitEvery)! or [`_.chunk`](https://lodash.com/docs/#chunk)

File: common/v2/services/ApiService/TokenInfo/TokenInfo.ts [hunk 2]
```diff
@@ -21,14 +22,37 @@ export default class TokenInfoService {
...
+    );
 
     try {
-      const response = await this.service.get('', { params });
-      return response.data;
+      return await Promise.all(
+        paramsArray.map(async (params: any) => this.service.get('', { params }))
+      ).then((returnedArray: AxiosResponse[]) =>
+        _.union(...returnedArray.map(returnObject => returnObject.data))
+      );
```
Comment (vilsbole) [inline common/v2/services/ApiService/TokenInfo/TokenInfo.ts]: You can increase readability by simplifying the steps and the naming:
```
  public getTokensInfo = async (contractAddresses: string[]) => {
    // Split contract addresses for large sets of contracts to break up the calls into manageable chunks
    // Explain why you choose 20
    const queryParams = chunk(20, contractAddresses).map(createParams);

    try {
      return await Promise.all(
          queryParams.map(async (params: any) => this.service.get('', { params }))
        ).then((res: AxiosResponse[]) =>
          _.union(...res.map(r => r.data))
        );
    } catch (e) {
      throw e;
    }
  };

```
- Avoiding intermediate variables increases readability.
- In TS adding the type to the variable name is redundant and can increase cognitive load.
- You're welcome to use acronyms for standard apis.

File: common/v2/services/RatesProvider.tsx [hunk 3]
```diff
@@ -36,32 +84,64 @@ export function RatesProvider({ children }: { children: React.ReactNode }) {
...
-  }, [rawAccounts, currentTickers.length]); // only update if an account has been added or removed from LocalStorage or if assets have been updated.
+      const terminateWorker = () => {
+        worker.stop();
+        worker.close();
+      };
+
+      // Stop old worker
+      terminateWorker();
+      // Create new worker.
+      createWorker();
```
Comment (vilsbole) [inline common/v2/services/RatesProvider.tsx]: The rules for stopping and starting the worker are quite confusing.
Can you share the conditions you identified?

File: common/v2/services/ApiService/TokenInfo/TokenInfo.ts [hunk 4]
```diff
@@ -21,14 +22,23 @@ export default class TokenInfoService {
...
-      const response = await this.service.get('', { params });
-      return response.data;
+      return Promise.all(
+        queryParams.map(async (params: any) => this.service.get('', { params }))
+      ).then((res: AxiosResponse[]) => _.union(...res.map(r => r.data)));
     } catch (e) {
       throw e;
     }
   };
 }
```
Comment (vilsbole) [inline common/v2/services/ApiService/TokenInfo/TokenInfo.ts]: If you return the Promise in a try/catch block, the catch block never executes (it's the catch of the Promise that will deal with any errors). Either:
- return the Promise without the try/catch
- `return await Promise.all(...)` inside a try/catch

https://jakearchibald.com/2017/await-vs-return-vs-return-await/

File: common/v2/services/RatesProvider.tsx [hunk 5]
```diff
@@ -1,30 +1,77 @@
...
 }
 
 const DEFAULT_FIAT_PAIRS = ['USD', 'EUR'] as TTicker[];
 const DEFAULT_FIAT_RATE = 0;
 const POLLING_INTERRVAL = 60000;
-const RATES_URL = 'https://proxy.mycryptoapi.com/cc/multi';
-const buildQueryUrl = (assets: TTicker[], currencies: TTicker[]) => `
-  ${RATES_URL}/?fsyms=${assets.join(',')}&tsyms=${currencies.join(',')}
+const ASSET_ID_MAPPING_URL =
+  'https://raw.githubusercontent.com/MyCryptoHQ/assets/master/assets/assets.json';
```
Comment (vilsbole) [inline common/v2/services/RatesProvider.tsx]: Hasn't the proxy been updated since?

File: common/v2/services/RatesProvider.tsx [hunk 6]
```diff
@@ -1,30 +1,77 @@
...
+const ASSET_RATES_URL = 'https://api.coingecko.com/api/v3/simple/price';
+const buildAssetQueryUrl = (assets: string[], currencies: string[]) => `
+  ${ASSET_RATES_URL}/?ids=${assets}&vs_currencies=${currencies}
 `;
 
+const fetchAssetMappingList = async (): Promise<AssetMappingListObject | any> => {
+  return fetch(ASSET_ID_MAPPING_URL, {
```
Comment (vilsbole) [inline common/v2/services/RatesProvider.tsx]: Prefer using the `ApiService` over `fetch`. It allows for a single place to deal with
- request headers
- failed calls
- response parsing
- throttling and retries.

File: common/v2/services/RatesProvider.tsx [hunk 7]
```diff
@@ -1,30 +1,77 @@
...
+    [uuid for coinGeckoId "ethereum"]: { "usd": 123.45, "eur": 234.56 }
+  } */
+
+  const updateRateObj = (acc: any, curValue: TTicker): IRates => {
+    const data: any = Object.keys(assetMap).find(uuid => assetMap[uuid].coinGeckoId === curValue);
+    acc[data] = rates[curValue];
+    return acc;
+  };
+
+  return Object.keys(rates).reduce(updateRateObj, {} as IRates);
```
Comment (vilsbole) [inline common/v2/services/RatesProvider.tsx]: Yummy!

File: common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx [hunk 8]
```diff
@@ -27,33 +28,35 @@ export function TokenPanel() {
...
       const tokenTotals = totals(selectedAccounts);
-
-      // Fetch details for all tokens of selected accounts
-      const tokensDetails: any[] = await TokenInfoService.instance.getTokensInfo(
+      const detailedTokenData: any[] = await TokenInfoService.instance.getTokensInfo(
         tokenTotals.map(x => x.contractAddress!).filter(x => x)
       );
+      setTokenData(detailedTokenData);
+    };
+    fetchTokenDetails();
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx]: yeh, that's definitely a lot cleaner.

File: common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx [hunk 9]
```diff
@@ -27,33 +28,35 @@ export function TokenPanel() {
...
+            Object.assign(asset, {
+              rate: getAssetRate(asset) || 0,
+              details: tokenData.find(details => details.address === asset.contractAddress) || {}
+            })
+          ];
+    }, []);
 
-    getTokensWithDetails();
+    setAllTokens(tempTokens);
   }, [accounts, rates]);
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx]: > 4. Is this the only place in the app we need tokensWithDetails? Would it be useful to move it into StoreProvider?

**Pull request merged**